CRM 18651 - Add in Option Group and soft validation

seamuslee001 opened the pull request.

totten commented:
During installation and upgrade, we should probably be setting the `data_type` for a few OptionGroups, e.g.
  * Installation (e.g. patch `xml/templates/civicrm_data.tpl` and maybe run `regen.sh`; if `regen.sh` is failing, then manually patch `sql/civcrm_generated.mysql`)
  * Upgrade (e.g. `UPDATE civicrm_option_group SET data_type = 'Integer' WHERE name = 'event_type';`)

yashodha commented:
@seamuslee001 Could you please rebase your PR and also move the upgrade code to  4.7.10.mysql.tpl since that's the next release.

seamuslee001 commented:
@yashodha have done so now

yashodha commented on CRM/Upgrade/Incremental/sql/4.7.10.mysql.tpl:
```diff
@@ -12,3 +12,10 @@ ALTER TABLE civicrm_pledge_block ADD pledge_start_date varchar(64) NULL DEFAULT
+ADD `data_type` varchar(128) COLLATE utf8_unicode_ci DEFAULT NULL COMMENT 'Data Type of Option Group.';
+UPDATE civicrm_option_group SET `data_type` = 'Integer'
+WHERE name in ('activity_type', 'gender', 'payment_instrument', 'participant_role', 'event_type',
+  'activity_status',);
```
There is a syntax error - unwanted **,** after activity_status.

jitendrapurohit commented on CRM/Admin/Form/OptionGroup.php:
```diff
@@ -71,6 +71,8 @@ public function buildQuickForm() {
       CRM_Core_DAO::getAttribute('CRM_Core_DAO_OptionGroup', 'description')
     );
 
+    $this->addSelect('Data Type', CRM_Utils_Type::dataTypes(), TRUE);
```
This doesn't let Option Group Form to load. Form to add option group results into a fatal error.

>Cannot determine default entity. CRM_Admin_Form_OptionGroup should implement getDefaultEntity().

jitendrapurohit commented on CRM/Utils/Check/Component/OptionGroups.php:
```diff
@@ -0,0 +1,87 @@
+       __FUNCTION__,
+       ts('The Following Option Values contain value fields that do not match the Data Type of the Option Group</p>
+        <p><table><tbody><th>Option Group</th><th>Option Value</th></tbody><tbody>') .
+        explode('\n', $strings) . ts('</tbody></table></p>'),
```
I think this should be `implode` instead of `explode` ? As this doesn't display the problematic option group in the status page.

jitendrapurohit commented on CRM/Utils/Check/Component/OptionGroups.php:
```diff
@@ -0,0 +1,87 @@
+            $validate = CRM_Utils_Type::validate($value['value'], $optionGroup['data_type'], FALSE);
+            if (!$validate) {
+              $problemValues[] = array(
+                'group_name' => $optionGroup['label'],
```
`$optionGroup`  have `title` and not `label` as a column name. So I suspect this would not set anything to the `group_name` key.

AkA84 commented:
Why allowing the user to create an invalid event type at all? Couldn't the error be caught at form validation? I can still create an event type with value "foo" and get the "Unable to reach the server. Please refresh this page in your browser and try again." error when creating an event of that type

AkA84 commented on CRM/Admin/Form/Options.php:
```diff
@@ -376,6 +376,19 @@ public static function formRule($fields, $files, $self) {
       }
     }
 
+    $optionGroup = civicrm_api3('OptionGroup', 'get', array(
```
I would wrap this in a private method to make the code cleaner and more readable (kind of pseudo-code following):
```php
/**
 *
 *
 */
private function getOptionGroupDataType($groupName) {
  $optionGroup = civicrm_api3('OptionGroup', 'get', array(
    'sequential' => 1,
    'name' => $groupName,
  ));

  return CRM_Core_BAO_OptionGroup::getDataType($optionGroup['id']);
}

// ...

public static function formRule($fields, $files, $self) {
  $dataType = $self->getOptionGroupDataType($self->_gName);

  if ($dataType) {
    // ...
  }

 //..
}
```

AkA84 commented:
Anyway this PR still has some issues unsolved and the branch has conflicts that need to be fixed if we want to merge it

seamuslee001 commented:
@AkA84 Hi Allesandro, The idea of soft Validation came from @totten . I would be in favor of more harder failure but Tim suggested that we should do soft validation.

AkA84 commented:
hi @seamuslee001 , is this ready to be reviewed again?

totten commented:
+1 for soft-validation / warnings. I'm usually a fan of hard-validation / failures, but in this case soft-validation feels like it has better overall cost/benefit (although I don't have a strict formula to rationalize that).

seamuslee001 commented:
@AkA84 @totten @jitendrapurohit I've rebased this against master and moved upgrade code to 4.7.12.  I would appreciate another review of this with the hope this may get in for .12

jitendrapurohit commented on CRM/Admin/Form/Options.php:
```diff
@@ -376,10 +376,36 @@ public static function formRule($fields, $files, $self) {
       }
     }
 
+    $optionGroup = $this->getOptionGroupDataType($self->_gName);
```
I think `$this` call need to be changed to `self::` under static formRule() function.

jitendrapurohit commented:
Found a `CiviCRM_API3_Exception: "DB Error: no such field"` when editing an event type option after applying the changes.

seamuslee001 commented:
@jitendrapurohit do you have any more backtrace on the no such field error?

jitendrapurohit commented:
Sure. Here it is:

CiviCRM_API3_Exception: "DB Error: no such field"

>#0 /home/web/civi_master/civicrm/CRM/Admin/Form/Options.php(402): civicrm_api3("OptionGroup", "get", (Array:2))
#1 /home/web/civi_master/civicrm/CRM/Admin/Form/Options.php(379): CRM_Admin_Form_Options::getOptionGroupDataType("event_type")
#2 [internal function](): CRM_Admin_Form_Options::formRule((Array:9), (Array:0), Object(CRM_Admin_Form_Options))
#3 /home/web/civi_master/civicrm/packages/HTML/QuickForm.php(1594): call_user_func((Array:2), (Array:9), (Array:0), Object(CRM_Admin_Form_Options))
#4 /home/web/civi_master/civicrm/CRM/Core/Form.php(489): HTML_QuickForm->validate()

seamuslee001 commented:
@jitendrapurohit I found that if i ran regen.sh the api error goes away.

jitendrapurohit commented on CRM/Admin/Form/Options.php:
```diff
@@ -376,10 +376,36 @@ public static function formRule($fields, $files, $self) {
       }
     }
 
+    $optionGroup = self::getOptionGroupDataType($self->_gName);
```
Isn't this variable should be named as `$dataType` instead of `$dataType`.

seamuslee001 replied:
Thanks

jitendrapurohit commented:
jenkins, test this please

jitendrapurohit commented:
@seamuslee001 Can you please rebase and update this as it has gone stale.

seamuslee001 commented:
@jitendrapurohit Done

jitendrapurohit commented on CRM/Upgrade/Incremental/php/FourSeven.php:
```diff
@@ -828,4 +838,17 @@ public static function alterIndexAndTypeForImageURL() {
+   */
+  public static function addDataTypeColumnToOptionGroupTable() {
+    if (!CRM_Core_BAO_SchemaHandler::checkIfFieldExists('civicrm_option_group', 'data_type')) {
+      CRM_Core_DAO::executeQuery("ALTER TABLE `civicrm_option_group` ADD COLUMN `data_type` varchar(128) COLLATE utf8_unicode_ci DEFAULT NULL comment 'Data Type of Option Group.'");
```
There's an error on upgrade at this point. Screenshot -
![multi_upgrade_error](https://cloud.githubusercontent.com/assets/5929648/18309988/bbd16532-7519-11e6-88b6-b70d84fed631.png)

jitendrapurohit commented on CRM/Utils/Check/Component/OptionGroups.php:
```diff
@@ -0,0 +1,87 @@
+       __FUNCTION__,
+       ts('The Following Option Values contain value fields that do not match the Data Type of the Option Group</p>
+        <p><table><tbody><th>Option Group</th><th>Option Value</th></tbody><tbody>') .
+        implode('\n', $strings) . ts('</tbody></table></p>'),
```
This seems to affect the UI(append '\n' for every problematic entry). I see that imploding with empty string fixes it.
![string_error](https://cloud.githubusercontent.com/assets/5929648/18310183/a07f81d2-751a-11e6-8c82-b8470ecee786.png)

seamuslee001 commented:
@jitendrapurohit thanks for that I think have fixed both issues can you take a look?

seamuslee001 commented:
Jenkins test this please

jitendrapurohit commented on CRM/Admin/Form/Options.php:
```diff
@@ -376,10 +376,36 @@ public static function formRule($fields, $files, $self) {
+    $optionGroup = civicrm_api3('OptionGroup', 'get', array(
+      'sequential' => 1,
+      'name' => $optionGroupName,
+    ));
```
Can we use `getFieldValue()` directly here to get the group id instead of an API call ? Something like -

    $optionGroupId = CRM_Core_DAO::getFieldValue('CRM_Core_DAO_OptionGroup', $optionGroupName, 'id', 'name');

jitendrapurohit commented on CRM/Utils/Check/Component/OptionGroups.php:
```diff
@@ -0,0 +1,87 @@
+        $values = civicrm_api3('OptionValue', 'get', array(
+          'sequential' => 1,
+          'option_group_id' => $optionGroup['name'],
+        ));
```
I think this will be only limited to a count of 25. Can we use a BAO function `getOptionValuesArray()` instead of an API call here too ?

    $values = CRM_Core_BAO_OptionValue::getOptionValuesArray(option_group_id);

jitendrapurohit commented:
According to [changes done here](https://github.com/civicrm/civicrm-core/blob/master/CRM/Admin/Form/Options.php#L165-L172), we don't build a value text field for activity_type. So, I get an invalid soft validation error on saving an `activity_type` option value.

seamuslee001 commented:
@jitendrapurohit thanks for those pickups, I will make those fixes and try and work out a solution for activity_type. Will ping you once i have pushed changes

seamuslee001 commented:
Jenkins test this please

seamuslee001 commented:
thanks @jitendrapurohit I have fixed up the code now as per your comments and tested locally

seamuslee001 commented:
@jitendrapurohit Jitendra would you be able to test this again?

jitendrapurohit commented:
@seamuslee001 Yes, I'll test this very soon :-)

jitendrapurohit commented:
I've tested this on installation/upgradation. Option values are soft validated and problematic values are shown on the status page correctly. From my side, this is ok to merge.

Thanks.

totten commented:
Jenkins reports one test failure which is pre-existing/not-relevant.
